pkg/kv/kvclient/kvcoord/dist_sender.go

```diff
@@ -596,6 +637,101 @@ func makeDistSenderRangeFeedMetrics() DistSenderRangeFeedMetrics {
+	var labels []string
+	for _, t := range locality.Tiers {
+		labels = append(labels, fmt.Sprintf("from_%s", t.Key))
+	}
+	for _, t := range locality.Tiers {
+		labels = append(labels, fmt.Sprintf("to_%s", t.Key))
+	}
+	m := &DistSenderKVInterceptorMetrics{
+		baseLocality: locality,
+		EstimatedReplicationBytes: aggmetric.NewCounter(
+			metaDistSenderKVInterceptorEstimatedReplicationBytes, labels...,
+		),
+	}
+	m.mu.cachedPathMetrics = make(map[networkPath]networkPathMetrics)
+	m.mu.pathMetrics = make(map[string]networkPathMetrics)
+	return m
+}
+
+// MetricStruct implements the metrics.Struct interface.
+func (m *DistSenderKVInterceptorMetrics) MetricStruct() {}
+
+// updateEstimatedReplicationBytes updates the EstimatedReplicationBytes metric
+// that represents the estimated replication bytes for the given KV replication
+// network path.
+func (m *DistSenderKVInterceptorMetrics) updateEstimatedReplicationBytes(
+	fromNodeID roachpb.NodeID,
+	fromLocality roachpb.Locality,
+	toNodeID roachpb.NodeID,
+	toLocality roachpb.Locality,
+	inc int64,
+) {
+	m.getEstimatedReplicationBytes(fromNodeID, fromLocality, toNodeID, toLocality).Inc(inc)
+}
+
+// getEstimatedReplicationBytes returns the metric that represents the estimated
+// replication bytes for the given network path.
+func (m *DistSenderKVInterceptorMetrics) getEstimatedReplicationBytes(
+	fromNodeID roachpb.NodeID,
+	fromLocality roachpb.Locality,
+	toNodeID roachpb.NodeID,
+	toLocality roachpb.Locality,
+) *aggmetric.Counter {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Check if we have cached values.
+	np := networkPath{fromNodeID: fromNodeID, toNodeID: toNodeID}
+	if cached, ok := m.mu.cachedPathMetrics[np]; ok {
+		return cached.EstimatedReplicationBytes
+	}
+
+	// Check if we've already initialized the metric. This is the case if a
+	// different network path results in the same normalized locality values.
+	labelValues := makeLocalityLabelValues(m.baseLocality, fromLocality, toLocality)
+	storeKey := strings.Join(labelValues, ",")
+	pm, ok := m.mu.pathMetrics[storeKey]
+	if !ok {
+		pm = networkPathMetrics{
+			EstimatedReplicationBytes: m.EstimatedReplicationBytes.AddChild(labelValues...),
+		}
+		m.mu.pathMetrics[storeKey] = pm
+	}
+	m.mu.cachedPathMetrics[np] = pm
+
+	return pm.EstimatedReplicationBytes
+}
+
+// makeLocalityLabelValues returns a list of label values which can be used for
+// network path metrics. This applies a mapping approach where all the values in
+// "from" and "to" localities are taken up until the first value that differs.
+// Remaining values will be padded as empty strings. Note that the keys must
+// exist in the given base locality. See test cases for more information.
+func makeLocalityLabelValues(baseLocality, fromLocality, toLocality roachpb.Locality) []string {
+	labelValues := make([]string, 2*len(baseLocality.Tiers))
+	for i := 0; i < len(baseLocality.Tiers); i++ {
+		if i >= len(fromLocality.Tiers) ||
```

Comment:
nit/consider: You can pull the bounds check into the loop. That makes the break condition easier to read.

```go
localityTiers := min(len(baseLocality.Tiers), len(fromLocality.Tiers), len(toLocality.Tiers))
for i := 0; i < localityTiers; i++ {
    base := baseLocality.Tiers[i].Key
    if fromLocality.Tiers[i].Key != base || toLocality.Tiers[i].Key != base {
		break
	}
    ...
}
```